drivers: sensor: Add support for grow_r502a fingerprint sensor

DineshDK03 opened the pull request:
Added driver and other support needed for r502a fingerprint sensor 

Signed-off-by: Dinesh Kumar K <dinesh@linumiz.com>

tejlmand commented on drivers/sensor/grow_r502a/Kconfig:
```diff
@@ -0,0 +1,59 @@
+
+config GROW_R502A_GPIO_POWER
+	bool "GROW_R502A sensor VCC and VT GPIO"
+	default n
```
`default n` is not needed. `n` is default if not stated.
https://docs.zephyrproject.org/latest/guides/build/kconfig/tips.html#redundant-defaults

```suggestion
```

DineshDK03 replied:
thanks for the review, fixed!

tejlmand commented on samples/sensor/grow_r502a/CMakeLists.txt:
```diff
@@ -0,0 +1,8 @@
+# SPDX-License-Identifier: Apache-2.0
+
+cmake_minimum_required(VERSION 3.13.1)
```
```suggestion
cmake_minimum_required(VERSION 3.20)
```

DineshDK03 replied:
fixed!

mbolivar-nordic commented on drivers/sensor/grow_r502a/grow_r502a.c:
```diff
@@ -0,0 +1,700 @@
+	}
+
+	ret = gpio_pin_configure_dt(&drv_data->act_gpios,
+				GPIO_OUTPUT_ACTIVE | drv_data->act_gpios.dt_flags);
```
You don't need to OR in the DT flags when you use gpio_pin_configure_dt. This comment applies elsewhere in the patch.

mbolivar-nordic commented on drivers/sensor/grow_r502a/grow_r502a.c:
```diff
@@ -0,0 +1,700 @@
+
+	if (drv_data->buffer[6] != FPS_ACK_PACKET) {
+		LOG_ERR("Error receiving ack packet 0x%X", drv_data->buffer[6]);
+		return -EBADMSG;
```
EBADMSG is used throughout this driver but that means "Invalid STREAMS message". I think these should all be EIO.

mbolivar-nordic commented on drivers/sensor/grow_r502a/grow_r502a.h:
```diff
@@ -0,0 +1,170 @@
+};
+
+struct grow_r502a_data {
+	const struct device *dev;
+	char *uart_bus;
+
+	struct gpio_dt_spec vin_gpios;
+	struct gpio_dt_spec act_gpios;
+
+#ifdef CONFIG_GROW_R502A_TRIGGER
+	const struct device *uart_dev;
+	struct gpio_dt_spec  int_gpios;
```
Please move all the devices and gpio_dt_spec values to a driver configuration structure to save RAM, since they can be const. Then you can remove the unncessary char* field.

mbolivar-nordic commented on samples/sensor/grow_r502a/README.txt:
```diff
@@ -0,0 +1,5 @@
+
+Description:
+
+A simple example using the grow_r502a fingerprint sensor to enroll, delete and match fingerprint.
```
Please use a real RST file like the other sensor samples.

mbolivar-nordic commented on samples/sensor/grow_r502a/src/main.c:
```diff
@@ -0,0 +1,106 @@
+	struct sensor_value val, del;
+	int ret;
+
+	const struct device *dev =  device_get_binding(FPS_DEV_NAME);
```
Please use DEVICE_DT_GET_ANY or DEVICE_DT_GET_ONE. Make sure to read the documentation for the required error handling steps.

mbolivar-nordic commented on samples/sensor/grow_r502a/src/main.c:
```diff
@@ -0,0 +1,106 @@
+#include <drivers/uart.h>
+#include <drivers/gpio.h>
+
+#define FPS_DEV_NAME	DT_LABEL(DT_INST(0, hzgrow_grow_r502a))
```
Please don't use DT_INST in samples; that can be avoided with DEVICE_DT_GET_... as described elsewhere in my review comments. The device structure has a `name` field you can use for this.

DineshDK03 replied:
thanks for the reviews, Fixed this and all the above reviews

gmarull commented on drivers/sensor/grow_r502a/Kconfig:
```diff
@@ -0,0 +1,58 @@
+	help
+	  Stack size of thread used by the driver to handle interrupts.
+
+config GROW_R502A_GPIO_POWER
+	bool "GROW_R502A sensor VCC and VT GPIO"
+	help
+	  enable high voltage to vin_gpios and act_gpios.
```
help can be improved here. `Enable control of {...}. This allows {...}.`

gmarull commented on drivers/sensor/grow_r502a/grow_r502a.c:
```diff
@@ -0,0 +1,713 @@
+	pkg_len = len + FPS_CHECKSUM_LEN;
+	check_sum = pkg_len + pkg_type;
+
+	buffer[0] = FPS_STARTCODE >> 8;
+	buffer[1] = FPS_STARTCODE & 0xFF;
+	buffer[2] = (uint8_t)(FPS_DEFAULT_ADDRESS >> 24);
+	buffer[3] = (uint8_t)(FPS_DEFAULT_ADDRESS >> 16);
+	buffer[4] = (uint8_t)(FPS_DEFAULT_ADDRESS >> 8);
+	buffer[5] = (uint8_t)(FPS_DEFAULT_ADDRESS & 0xFF);
+	buffer[6] = pkg_type;
+	buffer[7] = pkg_len >> 8;
+	buffer[8] = pkg_len & 0xFF;
```
please use `<sys/byteorder.h>` APIs

gmarull commented on drivers/sensor/grow_r502a/grow_r502a.c:
```diff
@@ -0,0 +1,713 @@
+	}
+
+	if (drv_data->buffer[9] == FPS_OK) {
+		LOG_INF("FPS LED ON");
```
`LOG_DBG`

gmarull commented on drivers/sensor/grow_r502a/grow_r502a.c:
```diff
@@ -0,0 +1,713 @@
+		return -EIO;
+	}
+
+	if (drv_data->buffer[9] == FPS_OK) {
```
this and others: if certain buffer indexes contain named fields, do something like this:

```c
#define MY_FIELD 9

if (drv_data->buffer[MY_FIELD] == FPS_OK) {
```

gmarull commented on drivers/sensor/grow_r502a/grow_r502a.c:
```diff
@@ -0,0 +1,713 @@
+static uint8_t fps_verify_password(const struct device *dev)
+{
+	struct grow_r502a_data *drv_data = dev->data;
+	uint8_t data[5] = {FPS_VERIFYPASSWORD, (uint8_t)(FPS_DEFAULT_PASSWORD >> 24),
+		(uint8_t)(FPS_DEFAULT_PASSWORD >> 16), (uint8_t)(FPS_DEFAULT_PASSWORD >> 8),
+		(uint8_t)(FPS_DEFAULT_PASSWORD & 0xFF)};
```
ditto, byteorder.h

gmarull commented on drivers/sensor/grow_r502a/grow_r502a.c:
```diff
@@ -0,0 +1,713 @@
+
+	ret = fps_image2Tz(dev, 1);
+	if (ret == FPS_OK) {
+		LOG_INF("Image converted");
```
`DBG`

gmarull commented on drivers/sensor/grow_r502a/grow_r502a.c:
```diff
@@ -0,0 +1,713 @@
+	return drv_data->buffer[9];
+}
+
+static uint8_t fps_image2Tz(const struct device *dev, uint8_t slot)
```
`use_lowercase_with_underscores_naming`

gmarull commented on drivers/sensor/grow_r502a/grow_r502a.c:
```diff
@@ -0,0 +1,713 @@
+				uint16_t *finger_id, uint16_t *matching_score)
+{
+	struct grow_r502a_data *drv_data = dev->data;
+	uint8_t data[6] = {FPS_SEARCH, slot, 0x00, 0x00, (uint8_t)(FPS_DEFAULT_CAPACITY >> 8),
+				(uint8_t)(FPS_DEFAULT_CAPACITY & 0xFF)};
```
another suggestion: seems that code can be formatted better. See https://github.com/zephyrproject-rtos/zephyr/pull/41307

gmarull commented on drivers/sensor/grow_r502a/grow_r502a.c:
```diff
@@ -0,0 +1,713 @@
+#ifdef CONFIG_GROW_R502A_GPIO_POWER
+	if (!device_is_ready(cfg->vin_gpios.port)) {
+		LOG_ERR("GPIO port %s not ready", cfg->vin_gpios.port->name);
+		return -EINVAL;
```
`-ENODEV`

gmarull commented on drivers/sensor/grow_r502a/grow_r502a.c:
```diff
@@ -0,0 +1,713 @@
+	}
+
+	gpio_pin_set_dt(&cfg->vin_gpios, 0);
+	k_sleep(K_MSEC(100));
```
magic number: please define

gmarull commented on drivers/sensor/grow_r502a/grow_r502a.c:
```diff
@@ -0,0 +1,713 @@
+
+	if (!device_is_ready(cfg->act_gpios.port)) {
+		LOG_ERR("GPIO port %s not ready", cfg->act_gpios.port->name);
+		return -EINVAL;
```
`-ENODEV`

DineshDK03 replied:
Thanks for the review, Solved this and all other changes requested on previous comments.

DineshDK03 committed:
dts: vendor-prefixes: Add hzgrow prefix

we have hzgrow,r502a binding.

Signed-off-by: Dinesh Kumar K <dinesh@linumiz.com>

DineshDK03 committed:
drivers: sensor: Add support for grow_r502a fingerprint sensor

Add driver support for grow_r502a fingerprint sensor

Signed-off-by: Dinesh Kumar K <dinesh@linumiz.com>

DineshDK03 committed:
samples: sensor: Added sample for grow_r502a fingerprint sensor

Added sample application for grow_r502a fingerprint sensor

Signed-off-by: Dinesh Kumar K <dinesh@linumiz.com>

gmarull commented on drivers/sensor/grow_r502a/grow_r502a.c:
```diff
@@ -0,0 +1,697 @@
+		int64_t remaining = end - sys_clock_tick_get();
+
+		if (remaining <= 0) {
+			return -EIO;
```
`-ETIMEDOUT`

gmarull commented on drivers/sensor/grow_r502a/grow_r502a.c:
```diff
@@ -0,0 +1,697 @@
+		return -EIO;
+	}
+
+	*template_cnt = drv_data->buffer[FPS_HEADER_LEN + 1];
+	*template_cnt <<= 8;
+	*template_cnt |= drv_data->buffer[FPS_HEADER_LEN + 2];
```
`byteorder` API here?

```
*template_cnt = sys_get_be16(&drv_data->buffer[FPS_HEADER_LEN + 1]);
```

gmarull commented on drivers/sensor/grow_r502a/grow_r502a.c:
```diff
@@ -0,0 +1,697 @@
+			return ret;
+		}
+		LOG_ERR("Error getting image 0x%X", ret);
+		return -EIO;
```
please propagate error codes

gmarull commented on drivers/sensor/grow_r502a/grow_r502a.c:
```diff
@@ -0,0 +1,697 @@
+	return drv_data->buffer[FPS_CC_IDX];
+}
+
+static uint8_t fps_image_to_char(const struct device *dev, uint8_t slot)
+{
+	struct grow_r502a_data *drv_data = dev->data;
+	uint8_t data[2] = { FPS_IMAGE2TZ, slot };
+
+	send_packet(dev, FPS_COMMAND_PACKET, data, sizeof(data));
+
+	int ret = receive_data(dev, drv_data->buffer);
+
+	if (ret < 0) {
+		return ret;
+	}
+
+	if (drv_data->buffer[FPS_PID_IDX] != FPS_ACK_PACKET) {
+		LOG_ERR("Error receiving ack packet 0x%X", drv_data->buffer[FPS_PID_IDX]);
+		return -EIO;
+	}
+
+	return drv_data->buffer[FPS_CC_IDX];
```
this is mixing error return codes (`int`) with `uint8_t` data, which is wrong.

gmarull commented on drivers/sensor/grow_r502a/grow_r502a.c:
```diff
@@ -0,0 +1,697 @@
+
+	if (rc < 0) {
+		LOG_ERR("Failed to initialize interrupt!");
+		return -EIO;
```
this and others: please propagate error codes.

gmarull commented on drivers/sensor/grow_r502a/grow_r502a.c:
```diff
@@ -0,0 +1,697 @@
+	const struct grow_r502a_config *cfg = dev->config;
+	int ret;
+
+	ret = fps_get_template_count(cfg->bus, &drv_data->template_count);
+	if (ret == FPS_OK) {
+		LOG_DBG("Read success");
+	} else {
+		LOG_ERR("FPS template count get error 0x%X", ret);
+		return -EIO;
+	}
```
suggestion, looks like there's a lot of boilerplate here, why not just

```
return fps_get_template_count(cfg->bus, &drv_data->template_count);
```

and let the application log status?

DineshDK03 replied:
thanks for the review, resolved this and all the above mentioned issues.

DineshDK03 replied:
@gmarull Kindly re-review for any issues.Thank You!

mbolivar-nordic commented on drivers/sensor/grow_r502a/Kconfig:
```diff
@@ -0,0 +1,58 @@
+config GROW_R502A_GPIO_POWER
+	bool "GROW_R502A sensor VCC and VT GPIO"
+	help
+	  Enable control of vin_gpios and act_gpios.
```
```suggestion
	  Enable control of vin-gpios and act-gpios.
```

mbolivar-nordic commented on tests/drivers/build_all/sensor/app.overlay:
```diff
@@ -123,7 +123,7 @@
 			status = "okay";
 
-			#include "uart.dtsi"
+                        #include "uart.dtsi"
```
Unnecessary change; please remove

mbolivar-nordic commented on tests/drivers/build_all/sensor/app.overlay:
```diff
@@ -133,6 +133,7 @@
 			dio-gpios = <&test_gpio 0 0>;
 			/* dht22; */
 		};
+
```
Unnecessary change; please remove

mbolivar-nordic commented on samples/sensor/grow_r502a/README.rst:
```diff
@@ -0,0 +1,56 @@
+Overview
+********
+
+This sensor sample shows template count inside fingerprint sensor and do the following,
+Finger enrollment ,fingerprint Matching and deleting. Sensor LED will blink accordingly.
+
+enrollment and matching should work only when trigger mode is set.
```
Please fix this up.

mbolivar-nordic replied:
I'm sorry, but I still can't parse this section; can you please try to fix the grammar?

DineshDK03 replied:
This is the elloborate content I am trying to say. are you expecting like this or do I have to explain the functions of the sensor here instead of explaining the sample working?

```
This sensor sample shows the number of fingerprints (template count) inside the sensor.
This sample adds a new fingerprint (enrollment) when SENSOR_ATTR_ADD is set, matches
Your placed finger with an existing fingerprint inside sensor and shows the matching ID and
its confidence when SENSOR_ATTR_FIND is get, deletes stored fingerprint using its 
ID when SENSOR_ATTR_DEL is set.

Tri-colour LED around fingerprint sensor hardware will blink or flash for each
Attributes,for successful addition and deletion of fingerprint blue LED will blink
Three times, for successful matching purple LED will flash. for all other error cases
Red LED will blink.

Fingeprint adding and matching functionalities only works when SENSOR_TRIG_TOUCH is set.
```

DineshDK03 replied:
@mbolivar-nordic  ping!

mbolivar-nordic replied:
I'm saying there are typos and grammar problems there, and in the comment here: https://github.com/zephyrproject-rtos/zephyr/pull/40394#discussion_r826601128. They are making it hard for me t understand this. Can you please fix the English?

DineshDK03 replied:
```
This sample has the below functionalities:
1. Shows the number of fingerprints stored in the sensor.
2. When SENSOR_ATTR_ADD is set then it adds a new fingerprint to the sensor.
3. When SENSOR_ATTR_FIND is set then it tries to find a match for the input fingerprint. On finding a match it returns ID and confidence.
4. When SENSOR_ATTR_DEL is set then it deletes a fingerprint from the sensor.
Note: Fingerprint add & delete functionalities work only when SENSOR_TRIG_TOUCH is set.
Tricolored LED in the sensor hardware will, flash on the below conditions:
1. On successful addition or deletion of fingerprint it will flash in blue three times.
2. On finding a match for the input fingerprint it will flash in purple.
3. In all other cases it will flash in red.
```

@mbolivar-nordic  Hope this overview is understandable and also having no grammatical mistakes.

mbolivar-nordic commented on samples/sensor/grow_r502a/README.rst:
```diff
@@ -0,0 +1,56 @@
+    Matched ID : 2
+    confidence : 169
+
+<repeats endlessly>
```
Please put this in the code block or turn it into a complete sentence instead

DineshDK03 replied:
Thanks for the review, resolved this and all the above mentioned comments.

mbolivar-nordic reviewed (commented):
Thank you for the patch. I noticed a few minor things.

mbolivar-nordic commented:
I'm still having trouble understanding what you're trying to say in the README -- all my other comments are addressed. Thank you.

MaureenHelm commented on include/drivers/sensor.h:
```diff
@@ -177,6 +177,9 @@ enum sensor_channel {
 	/** Desired charging current in mA */
 	SENSOR_CHAN_GAUGE_DESIRED_CHARGING_CURRENT,
 
+	/** Fingerprint Enrolling and searching channel */
```
units?

DineshDK03 replied:
It is fingerprint sensor, it will give template(fingerprint) counts and we can pass ID number for storing fingerprint in the sensor . I hope latest modified comment is enough.

DineshDK03 commented:
@mbolivar-nordic  , @MaureenHelm  Hope I fixed all the changes requested from both of you. please revisit PR and review for any changes needed further. Thank you.

DineshDK03 commented:
@mbolivar-nordic  and @MaureenHelm  ping!

DineshDK03 committed:
tests: drivers: build_all: sensor: Add grow_r502a fingerprint sensor

Included GROW_R502A fingerprint sensor in the build_all test to test
grow_r502a driver

Signed-off-by: Dinesh Kumar K <dinesh@linumiz.com>

DineshDK03 commented:
Added a few features and modified the driver with interrupt-driven APIs. Kindly review these modified codes and suggest to me a better way to get this driver merged as quickly as possible @mbolivar-nordic  @MaureenHelm

galak commented on drivers/sensor/grow_r502a/Kconfig:
```diff
@@ -0,0 +1,59 @@
+# SPDX-License-Identifier: Apache-2.0
+
+menuconfig GROW_R502A
+	bool "GROW_R502A Fingerprint Sensor"
```
Update to use new Kconfig.dts (generated) pattern.  See other drivers/sensors/*/Kconfig for examples:

```
default y
depends on DT_HAS_HZGROW_GROW_R502A_ENABLED
```

DineshDK03 replied:
fixed.

galak commented on dts/bindings/sensor/hzgrow,grow-r502a.yaml:
```diff
@@ -0,0 +1,33 @@
+# Copyright (c) 2021, Linumiz
+# SPDX-License-Identifier: Apache-2.0
+
+description: |
```
`description: HZ-Grow R502A Fingerprint sensor.`

DineshDK03 replied:
fixed

galak commented on dts/bindings/sensor/hzgrow,grow-r502a.yaml:
```diff
@@ -0,0 +1,33 @@
+description: |
+        HZ-Grow GROW_R502A Fingerprint sensor.
+
+compatible: "hzgrow,grow-r502a"
```
Make this `hzgrow,r502a` (Update filename to match)

DineshDK03 replied:
Then renaming from grow_r502a to r502a in all the file names also makes sense right?

galak replied:
Any where that you had duplicate HZGROW, GROW R502a.  Its the duplicate that needs cleaning up.

DineshDK03 replied:
Okay, Fixed!

galak commented on tests/drivers/build_all/sensor/uart.dtsi:
```diff
@@ -15,3 +16,9 @@ test_uart_mhz19b {
+
+test_uart_grow_r502a: grow_r502a {
+	compatible = "hzgrow,grow-r502a";
+	label = "GROW_R502A";
```
remove `label` not required.

DineshDK03 replied:
fixed

galak commented on samples/sensor/grow_r502a/sample.yaml:
```diff
@@ -0,0 +1,14 @@
+    tags: sensors
+    harness: sensor
+    build_only: true
+    depends_on: serial arduino_serial gpio
```
why does this depend on `arduino_serial`?

DineshDK03 replied:
I will remove it, as it already depends on serial.

galak commented on samples/sensor/grow_r502a/prj.conf:
```diff
@@ -0,0 +1,7 @@
+CONFIG_UART_INTERRUPT_DRIVEN=y
+
+CONFIG_SENSOR=y
+CONFIG_GROW_R502A=y
```
`CONFIG_GROW_R502A=y` not needed will get set by Kconfig & dts automatically.

DineshDK03 replied:
fixed

galak commented on samples/sensor/grow_r502a/boards/nrf52840dk_nrf52840.overlay:
```diff
@@ -0,0 +1,15 @@
+        status = "okay";
+        grow_r502a {
+                compatible = "hzgrow,grow-r502a";
+                label = "GROW_R502A";
```
remove label property.

DineshDK03 replied:
fixed

galak commented on samples/sensor/grow_r502a/boards/frdm_k64f.overlay:
```diff
@@ -0,0 +1,15 @@
+        status = "okay";
+        grow_r502a {
+                compatible = "hzgrow,grow-r502a";
+                label = "GROW_R502A";
```
remove label property.

DineshDK03 replied:
fixed

galak commented on samples/sensor/grow_r502a/boards/esp32.overlay:
```diff
@@ -0,0 +1,24 @@
+
+	grow_r502a {
+		compatible = "hzgrow,grow-r502a";
+		label = "grow_r502a";
```
remove label property

DineshDK03 replied:
fixed

galak reviewed (changes requested):
A few Kconfig and dts related changes requested.

DineshDK03 commented:
@galak Fixed all the changes requested from you. please have a look, Thanks.

galak commented on drivers/sensor/r502a/Kconfig:
```diff
@@ -0,0 +1,62 @@
+# Copyright (c) 2021 Linumiz
+# SPDX-License-Identifier: Apache-2.0
+
+menuconfig R502A
```
make this `GROW_R502A`

DineshDK03 replied:
But, If the compatible name changes to hzgrow,r502a then it is better to have r502a as driver and config name right?

DineshDK03 replied:
@galak  If you think otherwise, I am ready to push the changes you asked for. Do I have to rename driver file names and kconfig entries to grow_r502a and GROW_R502A or keep the current changes as r502a alone?

DineshDK03 replied:
@galak Renamed as per your comments.

DineshDK03 replied:
@galak ping, kindly re-review the changes and help me mainlining this driver.

galak reviewed (dismissed):
sorry if I wasn't clear, but a few minor things w/name:

1. Rename Kconfig options to be `GROW_R502A`. (R502A_THREAD_PRIORITY -> GROW_R502A_THREAD_PRIORITY).
2. rename dir from drivers/sensor/r502a/ to drivers/sensor/grow_r502a/

galak commented on dts/bindings/sensor/hzgrow,r502a.yaml:
```diff
@@ -0,0 +1,32 @@
+    address:
+      type: int
+      required: false
+      default: 0xffffffff
```
I'd prefer to remove `default` and just use DT_INST_PROP_OR() in the driver.

DineshDK03 replied:
I fixed this, and now I am getting this address via reg property by including uart-controller.yaml as per @gmarull suggestion.

gmarull commented on include/zephyr/drivers/sensor.h:
```diff
@@ -313,6 +320,27 @@ enum sensor_attribute {
 	/** Alert threshold or alert enable/disable */
 	SENSOR_ATTR_ALERT,
 
+	/**
+	 * To enroll fingerprint image as template
+	 */
+	SENSOR_ATTR_ADD,
+	/**
+	 * To search fingerprint template
+	 */
+	SENSOR_ATTR_FIND,
+	/**
+	 * To delete fingerprint template or to empty template database
+	 */
+	SENSOR_ATTR_DEL,
+
+	/**
+	 * To get free Fingerprint template index for enrolling
+	 */
+	SENSOR_ATTR_FREE_IDX,
+	/**
+	 * To empty Fingerprint library
+	 */
+	SENSOR_ATTR_EMPTY,
```
how do I know these are attributes for a fingerprint "SENSOR_ATTR_ADD"? I'd suggest using custom driver attributes here.

DineshDK03 replied:
understood and added generic attributes SENSOR_ATTR_RECORD_ which will be common to any sensor which is having internal storage.

gmarull commented on drivers/sensor/grow_r502a/grow_r502a.h:
```diff
@@ -0,0 +1,192 @@
+ *--------------------------------------------------------------------------------
+ */
+
+#include <zephyr/sys/ring_buffer.h>
```
this header is not self-contained, also please, only put things that are shared between modules in internal headers.

DineshDK03 replied:
removed it in the latest changes as I am not using ring buffer

gmarull commented on dts/bindings/sensor/hzgrow,r502a.yaml:
```diff
@@ -0,0 +1,32 @@
+include: uart-device.yaml
+
+properties:
+    address:
+      type: int
+      required: false
+      default: 0xffffffff
```
Even though uart isn't strictly a bus, it can be depending on the physical configuration. In Zephyr, `uart-controller.yaml` defines UART as a bus, so you can add devices connected to the UART as child nodes and use the standard `reg` property.

DineshDK03 replied:
understood and fixed.

gmarull commented on samples/sensor/grow_r502a/boards/esp32.overlay:
```diff
@@ -0,0 +1,23 @@
+	pinctrl-0 = <&uart2_default>;
+	pinctrl-names = "default";
+
+	grow_r502a {
+		compatible = "hzgrow,r502a";
+		int-gpios = <&gpio0 27 GPIO_ACTIVE_LOW>;
+	};
```
this is effectively using uart as a bus, see my comment above.

DineshDK03 replied:
yes, fixed

gmarull commented on samples/sensor/grow_r502a/boards/frdm_k64f.overlay:
```diff
@@ -0,0 +1,14 @@
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+&uart3 {
+        current-speed = <57600>;
+        status = "okay";
+        grow_r502a {
+                compatible = "hzgrow,r502a";
+                int-gpios = <&gpioc 12 GPIO_ACTIVE_LOW>;
+        };
+};
```
please use tabs, no spaces.

DineshDK03 replied:
removed this overlay, as one overlay is enough for a sample.

gmarull commented on samples/sensor/grow_r502a/sample.yaml:
```diff
@@ -0,0 +1,14 @@
+tests:
+  sample.sensor.grow_r502a:
+    tags: sensors
+    harness: sensor
+    build_only: true
+    depends_on: serial gpio
```
?

DineshDK03 replied:
this sensor depends on serial only understood and fixed

gmarull commented on samples/sensor/grow_r502a/src/main.c:
```diff
@@ -0,0 +1,121 @@
+	static struct sensor_value del, fid_get;
+	int ret;
+
+	const struct device *dev =  DEVICE_DT_GET_ANY(hzgrow_r502a);
+
+	if (dev ==  NULL) {
+		printk("Error: no device found\n");
+		return;
+	}
```
suggestion: use `DEVICE_DT_GET_ONE`

DineshDK03 replied:
Okay, used this in latest changes

ssekar15 commented on drivers/sensor/grow_r502a/grow_r502a.c:
```diff
@@ -0,0 +1,753 @@
+
+	grow_r502a_uart_flush(cfg->dev);
+
+	ring_buf_init(&drv_data->ringbuf, sizeof(drv_data->ring_buffer), drv_data->ring_buffer);
```
Ringbuffer is not needed. Producer and consumer are same, only one element at any instance in the ring buffer, so please use simple buffer.

DineshDK03 replied:
Thanks for the review, it is addressed in the latest changes

ssekar15 commented on drivers/sensor/grow_r502a/grow_r502a.c:
```diff
@@ -0,0 +1,753 @@
+	return 0;
+}
+
+static int fps_enroll(const struct device *dev, const struct sensor_value *val)
```
All these sequence should be atomic, because the data manipulation of each command are stored in temporary buffer inside chip. Please use locks

DineshDK03 replied:
fixed

ssekar15 commented on drivers/sensor/grow_r502a/grow_r502a.c:
```diff
@@ -0,0 +1,753 @@
+	drv_data->buffer[FPS_PID_IDX] = pkg_type;
+	sys_put_be16(pkg_len, &drv_data->buffer[FPS_PKG_LEN_IDX]);
+	for (int i = FPS_HEADER_LEN; i < (len + FPS_HEADER_LEN); i++) {
+		drv_data->buffer[i] = pkg_data[i - FPS_HEADER_LEN];
```
all the buffer to send_packet api are internal to the driver, avoid copying

DineshDK03 replied:
Okay, this too addressed in the latest commits

ssekar15 commented on include/zephyr/drivers/sensor.h:
```diff
@@ -313,6 +320,27 @@ enum sensor_attribute {
+	/**
+	 * To enroll fingerprint image as template
+	 */
+	SENSOR_ATTR_ADD,
```
All these new attributes are not specific to fingerprint, any chipset support internal record/database maintenance. So please change comments accordingly and change naming as SENSOR_ATTR_RECORD_*

DineshDK03 replied:
okay, updated and fixed in the latest changes

DineshDK03 commented:
rebased to the latest,  requesting re-review from @MaureenHelm  @mbolivar-nordic @galak @gmarull

ssekar15 reviewed (dismissed):
Looks good for me.

DineshDK03 commented:
@mbolivar-nordic  @gmarull  all the previous reviews requested are addressed, kindly re-review and help me mainlining this driver, Thanks.

DineshDK03 commented on include/zephyr/drivers/sensor.h:
```diff
@@ -312,6 +319,16 @@ enum sensor_attribute {
 	SENSOR_ATTR_FEATURE_MASK,
 	/** Alert threshold or alert enable/disable */
 	SENSOR_ATTR_ALERT,
+	/** Add values to the sensor which are having record storage facility */
+	SENSOR_ATTR_RECORD_ADD,
+	/** To find requested data in record storage */
+	SENSOR_ATTR_RECORD_FIND,
+	/** To delete mentioned data from record storage */
+	SENSOR_ATTR_RECORD_DEL,
+	/** To get available position to store data on record storage */
+	SENSOR_ATTR_RECORD_FREE_IDX,
+	/** To empty the storage record*/
+	SENSOR_ATTR_RECORD_EMPTY,
```
These attributes are common for any sensors which are having internal storage/ database within its hardware itself.

MaureenHelm replied:
Can you point to some examples please?

DineshDK03 replied:
Apart from Hzgrow, Here are a few examples of the fingerprint sensor from other manufacturers that have internal record maintenance https://fingerprint-hardware-software.de/wp-content/uploads/2019/01/b1-upek-fs-bpd.pdf , https://cdn.sparkfun.com/assets/5/a/3/f/c/Module_User_Manual-17151-Capacitive_Fingerprint_Scanner_-_AS-108M.pdf

MaureenHelm replied:
I'd still like to keep these as driver-specific for now

DineshDK03 replied:
Okay, I moved them to the driver-specific header.

MaureenHelm commented on drivers/sensor/grow_r502a/grow_r502a.c:
```diff
@@ -0,0 +1,787 @@
+		goto unlock;
+	}
+
+	ret = -1;
```
Remove

MaureenHelm commented on drivers/sensor/grow_r502a/grow_r502a.c:
```diff
@@ -0,0 +1,787 @@
+		goto unlock;
+	}
+
+	ret = fps_image_to_char(dev, 2);
```
Please use self-documenting macros for these special values

MaureenHelm commented on drivers/sensor/grow_r502a/grow_r502a.c:
```diff
@@ -0,0 +1,787 @@
+		return -ENODEV;
+	}
+
+#ifdef CONFIG_GROW_R502A_GPIO_POWER
```
```suggestion
	if (IS_ENABLED(CONFIG_GROW_R502A_GPIO_POWER)) {
```

MaureenHelm commented on dts/bindings/sensor/hzgrow,r502a.yaml:
```diff
@@ -0,0 +1,28 @@
+
+compatible: "hzgrow,r502a"
+
+include: uart-controller.yaml
```
```suggestion
include: [sensor-device.yaml, uart-device.yaml]
```

MaureenHelm commented on tests/drivers/build_all/sensor/app.overlay:
```diff
@@ -117,6 +117,8 @@
 		test_uart: uart@55556666 {
 			compatible = "vnd,serial";
 			reg = <0x55556666 0x1000>;
+			#address-cells = <1>;
+			#size-cells = <0>;
```
This is wrong

MaureenHelm commented on tests/drivers/build_all/sensor/uart.dtsi:
```diff
@@ -15,3 +16,9 @@ test_uart_mhz19b {
 	compatible = "plantower,pms7003";
 };
+
+test_uart_grow_r502a: grow_r502a@ffffffff {
```
This shouldn't have an address

DineshDK03 replied:
https://github.com/zephyrproject-rtos/zephyr/pull/40394#discussion_r943217396  there is a previous suggestion to use reg property from `uart-controller.yaml` , looking at the yaml reg property comes from `base.yaml` there reg property is optional, so either way, using `uart-device.yaml `and uart-controller.yaml makes no significant difference. but the actual issue is, this fingerprint sensor has a 4byte address byte going into data package. the suggestion was to remove custom defined addr property in `hzgrow,r502a.yaml ` and to use reg property , I make the changes according to that suggestion.

Here, coming to this issue, I need to use #address-cells = <1>; and optional #size-cells = <0>; property inside uart parent node to make the sample compilable. looking at how to not include #*-cells properties inside uart parent node is like below:

```
test_uart_grow_r502a {
        #address-cells=<1>;
        #size-cells=<0>;
        grow_r502a@0 {
        	compatible = "hzgrow,grow-r502a";
        	label = "GROW_R502A";
        	reg=<0x0>;
        	int-gpios = <&test_gpio 0 0>;
        };
};
```
is it okay to use node like the above snippet? @MaureenHelm

MaureenHelm commented:
Please fix the failed twister checks

DineshDK03 commented:
@MaureenHelm   fixed twister failures, pinging for review.

DineshDK03 commented:
@mbolivar-nordic  @gmarull  please review and help me merge this PR.

DineshDK03 commented:
@mbolivar-nordic  @MaureenHelm  @gmarull  kindly revisit.

MaureenHelm commented on include/zephyr/drivers/sensor/grow_r502a.h:
```diff
@@ -0,0 +1,43 @@
+
+enum sensor_attribute_grow_r502a {
+	/** Add values to the sensor which are having record storage facility */
+	SENSOR_ATTR_RECORD_ADD = SENSOR_ATTR_PRIV_START,
+	/** To find requested data in record storage */
+	SENSOR_ATTR_RECORD_FIND,
+	/** To delete mentioned data from record storage */
+	SENSOR_ATTR_RECORD_DEL,
+	/** To get available position to store data on record storage */
+	SENSOR_ATTR_RECORD_FREE_IDX,
+	/** To empty the storage record*/
+	SENSOR_ATTR_RECORD_EMPTY,
```
```suggestion
	SENSOR_ATTR_R502A_RECORD_ADD = SENSOR_ATTR_PRIV_START,
	/** To find requested data in record storage */
	SENSOR_ATTR_R502A_RECORD_FIND,
	/** To delete mentioned data from record storage */
	SENSOR_ATTR_R502A_RECORD_DEL,
	/** To get available position to store data on record storage */
	SENSOR_ATTR_R502A_RECORD_FREE_IDX,
	/** To empty the storage record*/
	SENSOR_ATTR_R502A_RECORD_EMPTY,
```

DineshDK03 replied:
fixed and did a rebase.

MaureenHelm commented on tests/drivers/build_all/sensor/uart.dtsi:
```diff
@@ -15,3 +16,13 @@ test_uart_mhz19b {
 };
+
+test_uart_grow_r502a {
+	#address-cells=<1>;
+	#size-cells=<0>;
+	grow_r502a@0 {
```
Why does this have an address?

DineshDK03 replied:
basically, this fingerprint sensor creates an RX packet based on the address we are going to pass, initially, it was passed to the driver via the address parameter from the fps node and then based on the PR suggestion from @gmarull  [here](https://github.com/zephyrproject-rtos/zephyr/pull/40394#discussion_r943217396)  address is given to driver via reg property. if you are having any conflict of opinion please let me know. thanks

DineshDK03 commented:
@mbolivar-nordic  @galak  @gmarull Ping, kindly re-visit the PR
